### PR TITLE
operator: remove empty JointConsensus steps and fix single voter demote bug

### PR DESCRIPTION
**Pull request by matchge-ca**
Signed-off-by: Hua Lu <hua.lu@pingcap.com>

### What problem does this PR solve?
Closes: #4362
Closes: #4444 

### What is changed and how it works?
- Modify condition to enable `useJointConsensus` when there is only single **non-demote** peer change
- Add logic in `execChangePeerV2` to prevent generate empty enter / leave JointConsensus steps
- Remove `allowDemote` and `DemoteFollower` step

### Check List

Tests

- Unit test

Related changes
- Need to cherry-pick to the release branch

### Release note

```release-note
- fix a bug that operate steps may contain unnecessary or empty JointConsensus steps in certain conditions 
- fix a bug when demoting single voter directly
```

### CODE REVIEW

File: server/schedule/operator/builder_test.go [hunk 1]
```diff
@@ -168,44 +167,49 @@ func (s *testBuilderSuite) TestPrepareBuild(c *C) {
 
 func (s *testBuilderSuite) TestBuild(c *C) {
 	type testCase struct {
-		allowDemote       bool
+		name              string
```
Comment (HunDunDM) [inline server/schedule/operator/builder_test.go:170]: Never used, cannot pass linter.

File: server/schedule/operator/builder.go [hunk 2]
```diff
@@ -676,35 +668,49 @@ func (b *Builder) execRemovePeer(peer *metapb.Peer) {
...
-		peer := b.toPromote[p]
-		step.PromoteLearners = append(step.PromoteLearners, PromoteLearner{ToStore: peer.GetStoreId(), PeerID: peer.GetId()})
-		b.currentPeers.Set(peer)
-	}
-	b.toPromote = newPeersMap()
+		// Transfer Leader
+		if needTransferLeader && b.originLeaderStoreID != b.targetLeaderStoreID {
+			b.execTransferLeader(b.targetLeaderStoreID)
+		}
+	} else {
```
Comment (HunDunDM) [inline server/schedule/operator/builder.go:678]:
```suggestion
		}
		return
	}
```

File: server/schedule/operator/builder.go [hunk 3]
```diff
@@ -372,27 +368,23 @@ func (b *Builder) prepareBuild() (string, error) {
...
-				if b.allowDemote {
-					b.toDemote.Set(n)
-				} else {
-					b.toRemove.Set(o)
-					// Need to add `b.toAdd.Set(n)` in the later targetPeers loop
-				}
+		if core.IsLearner(o) && !core.IsLearner(n) {
+			// learner -> voter
+			b.toPromote.Set(n)
+		} else if !core.IsLearner(o) && core.IsLearner(n) {
```
Comment (HunDunDM) [inline server/schedule/operator/builder.go:374]: `core.IsLearner` is called repeatedly.

**Review by HunDunDM — APPROVED** [review]

**Comment by HunDunDM** [comment]
The previous lgtm was a misoperation, sorry.

**Review by HunDunDM — CHANGES_REQUESTED** [review]
The previous lgtm was a misoperation, sorry.

File: server/schedule/operator/builder.go (hunk 2 above)
Comment (HunDunDM) [reply server/schedule/operator/builder.go:671]: The content in the `else` is too long. In fact, `len(b.toPromote)+len(b.toDemote) == 0` is only a special case, so you can `return` directly in the `if`, and no `else` is needed later.

File: server/schedule/operator/builder.go [hunk 4]
```diff
@@ -676,35 +661,48 @@ func (b *Builder) execRemovePeer(peer *metapb.Peer) {
 }
 
 func (b *Builder) execChangePeerV2(needEnter bool, needTransferLeader bool) {
-	// Enter
-	step := ChangePeerV2Enter{
-		PromoteLearners: make([]PromoteLearner, 0, len(b.toPromote)),
-		DemoteVoters:    make([]DemoteVoter, 0, len(b.toDemote)),
-	}
+	if len(b.toPromote)+len(b.toDemote) == 0 {
+		// No need to add empty enter / leave joint consensus step if no peer in toPromote and toDemote
```
Comment (HunDunDM) [inline server/schedule/operator/builder.go:665]:
```suggestion
		// No need to add empty enter / leave joint consensus step if no peer in `toPromote` and `toDemote`
```

**Review by HunDunDM — COMMENTED** [review]
rest LGTM

**Review by HunDunDM — APPROVED** [review]

**Comment by HunDunDM** [comment]
ref https://github.com/tikv/tikv/issues/11463

**Review by disksing — APPROVED** [review]

**Pull request merged**